hazelcast/src/test/java/com/hazelcast/executor/CompletableFutureTest.java

```diff
@@ -0,0 +1,134 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(QuickTest.class)
+public class CompletableFutureTest extends HazelcastTestSupport {
+
+    private ExecutionService es;
+    private CountDownLatch startLatch, doneLatch;
+    private ExecutorService executorService;
+    private AtomicReference<Object> ref1, ref2;
+
+    @Before
+    public void setUp() throws Exception {
+        HazelcastInstanceProxy proxy = (HazelcastInstanceProxy) createHazelcastInstance();
+        Field originalField = HazelcastInstanceProxy.class.getDeclaredField("original");
+        originalField.setAccessible(true);
+        HazelcastInstanceImpl hz = (HazelcastInstanceImpl) originalField.get(proxy);
+        NodeEngine nodeEngine = hz.node.nodeEngine;
+        es = nodeEngine.getExecutionService();
+        startLatch = new CountDownLatch(1);
+        executorService = Executors.newSingleThreadExecutor();
+        ref1 = new AtomicReference<Object>();
+        ref2 = new AtomicReference<Object>();
+    }
+
+    @After
+    public void tearDown() {
+        executorService.shutdown();
+    }
+
+    @Test
+    public void testPreregistered() throws Exception {
+        doneLatch = new CountDownLatch(1);
+        final ICompletableFuture<String> f = submitAwaitStartLatch();
+        f.andThen(setRefAndBumpDoneLatch(ref1));
+        startLatch.countDown();
+        assertOpenEventually(doneLatch);
+        assertEquals("success", ref1.get());
+    }
+
+    @Test
+    public void testMultiPreregistered() throws Exception {
+        doneLatch = new CountDownLatch(2);
+        final ICompletableFuture<String> f = submitAwaitStartLatch();
+        f.andThen(setRefAndBumpDoneLatch(ref1));
+        f.andThen(setRefAndBumpDoneLatch(ref2));
+        startLatch.countDown();
+        assertOpenEventually(doneLatch);
+        assertEquals("success", ref1.get());
+        assertEquals("success", ref2.get());
+    }
+
+    @Test
+    public void testPostregistered() throws Exception {
+        doneLatch = new CountDownLatch(1);
+        final ICompletableFuture<String> f = submitOpenStartLatch();
+        assertOpenEventually(startLatch);
+        f.andThen(setRefAndBumpDoneLatch(ref1));
+        assertOpenEventually(doneLatch);
+        assertEquals("success", ref1.get());
+    }
+
+    @Test
+    public void testMultiPostregistered() throws Exception {
+        doneLatch = new CountDownLatch(2);
+        final ICompletableFuture<String> f = submitOpenStartLatch();
+        assertOpenEventually(startLatch);
+        f.andThen(setRefAndBumpDoneLatch(ref1));
+        f.andThen(setRefAndBumpDoneLatch(ref2));
+        assertOpenEventually(doneLatch);
+        assertEquals("success", ref1.get());
+        assertEquals("success", ref2.get());
+    }
+
```

Comment:
Can you also add some method that check exceptional behavior? Is the exception propagated correctly? And what about interruption and cancellation?